# Object Oriented Pytorch 

### Introduction

In the previous lessons, we became familiar with tensors, linear layers, and activation functions in Pytorch to initialize a model and perform a forward pass of the data.  So far, things have been pretty similar to what we have seen in numpy.  We initialize some sample numbers for our weights and biases, and use functions as our activation layers.  

We did this through two functions: an `init_model` function and a `forward` function.  Because these two functions relate to the same neural network, we might imagine combining them into a single class.  That is what we'll accomplish in this lesson.

Later on, we'll see that combining these functions in a class, allows Pytorch to easily calculate our gradients when we eventually train our neural network.

Let's get started.

### Moving Functions to Objects

So far, we have written two functions to initialize a model and pass through our data.

In [85]:
import torch.nn as nn
import torch.nn.functional as F
def init_model():
    W1 = nn.Linear(28*28, 64)
    W2 = nn.Linear(64, 10)
    return {'W1': W1, 'W2': W2} 

model = init_model()

model

{'W1': Linear(in_features=784, out_features=64, bias=True),
 'W2': Linear(in_features=64, out_features=10, bias=True)}

In [86]:
def forward(X, model):
    W1, W2 = tuple(model.values())
    Z1 = W1(X)
    A1 = F.sigmoid(Z1)
    Z2 = W2(A1)
    A2 = F.softmax(Z2, dim = 1)
    return (Z1, A1, Z2, A2)

Now moving these two functions into a class really doesn't change our two functions too much.

In [88]:
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.W1 = nn.Linear(28*28, 64)
        self.W2 = nn.Linear(64, 10)
        
    def forward(self, x):
        A1 = torch.sigmoid(self.W1(x))
        return F.softmax(self.W2(A1), dim = 1) 

There are two main changes here.  First, we move our init_model function into the `__init__` function.  As we know from object orientation, this function will now be called when we initialize our neural network.

In [89]:
Net()

Net(
  (W1): Linear(in_features=784, out_features=64, bias=True)
  (W2): Linear(in_features=64, out_features=10, bias=True)
)

In [93]:
# ??nn.Module

Secondly, we have our class inherit from our `nn.Module` function.  This adds functionality to our class, which we'll see later.

Finally, we set up call the `__init__` function in the `nn.Module` class, through the `super().__init__()` init line.

That's it.  We've successfully setup a neural network in Pytorch.

### Making Use of Our Network

So this is the code we needed to translate our functions to a class in Pytorch.  Now let's see what we get.

In [11]:
from torch import nn
import torch.nn.functional as F
class Net(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(28*28, 64)
        self.fc2 = nn.Linear(64, 10)
        
    def forward(self, x):
        A1 = torch.sigmoid(self.fc1(x))
        return F.softmax(self.fc2(A1), dim = 1)

The first step is to initialize our neural network, and assign it to a variable.

In [12]:
net = Net()

Next let's pass some data through the network.

> First we initialize an instance with one row, 784 features.

In [16]:
import torch
x = torch.randn(1, 28*28)
# x

Then we pass through our data to our neural network.

In [14]:
net(x)

tensor([[0.0882, 0.0702, 0.1430, 0.0827, 0.0477, 0.1715, 0.1099, 0.1002, 0.1011,
         0.0855]], grad_fn=<SoftmaxBackward>)

We can see that passing through the data automatically calls the `forward` function in our network.

In [18]:
def forward(self, x):
    A1 = torch.sigmoid(self.fc1(x))
    return F.softmax(self.fc2(A1), dim = 1)

With that we've completed our hypothesis function in Pytorch.

### Resources

[Towards data science Pytorch Gradients](https://towardsdatascience.com/understanding-pytorch-with-an-example-a-step-by-step-tutorial-81fc5f8c4e8e)

[Pytorch viz](https://github.com/szagoruyko/pytorchviz)